rfcs/20231108-gpt-quantization/README.md

```diff
@@ -0,0 +1,136 @@
+# Proposal for GPT-Quantization (GPT-Q) support
+
+# 1. Introduction
+
+The motivation for GPT-Q[[#1]][1] is to reduce memory requirements for storage and execution.
+GPT-Q is an attempt to reduce model size without enforcing quantization of activation in Linear
+layers. The idea is to compress weights once after training is completed. During inference,
+weights are decompressed into f16/f32 data types and passed into f16/f32 Linear layers.
+The decompression happens in local device memory or in registers to reduce global memory
+size and bandwidth requirements.
+
+The quantization details:
+- The potential weights compression data types are integer 8, 4, 3 and even 2 bits.
+- The quantization is asymmetric (both scales and zero points are used).
+- The quantization scales and zero points are multidimensional and contain groups.
+
+## OpenVino support
+
+Recently OpenVino added support for the following receipts of weights compression to support GPT-Q:
+1. int8 weights compression:
+   - Scales are per-output channels, data type is f16/f32
+   - Zero points are per-output channels, data type is u8/s8
+2. int4 weights compression [[#2]][2] [[#3]][3]:
+   - Scales are per-output and per-input channels with groups along input channels, data type is f16/f32
+   - Zero points are per-output and per-input channels with groups along input channels, data type is u8/s8
+
+# 2. Implications to oneDNN
+
+There are 3 new quantization features that can't be supported by the current oneDNN API:
+- scales and zero points data type. Currently oneDNN defines scales as f32 data and zero points as s32 data.
+- scales and zero points groups support. Currently scales and zero points are set using a mask where each
+  bit represents if for a given dimension multiple or a single scale/zero point is used.
+- Weights quantization applied to primitives with non-integer activations. This part is covered by a separate RFC: [[#4]][4]
+
+# 3. Proposal
+
+## Scales and zero points data type support
+
+### Option a (recommended)
+
+Extend scales and zero points setting API with a parameter for data type. This additional information will allow kernels to generate
+the best code for a given data type.
+
+```cpp
+// C API
+status_t dnnl_primitive_attr_set_scales(
+        primitive_attr_t *attr, int arg, ..., dnnl_data_type_t data_type);
+
+// C++ API
+void set_scales_mask(int arg, ..., data_type_t data_type);
+
+// An example
+primitive_attr attr;
+attr.set_scales(DNNL_ARG_DST, ..., dt::f16);
+...
+auto dst_scales_md = memory::desc({1}, dt::f16, tag::x);
+auto dst_scales = memory(dst_scales_md, engine);
+...
+p.execute(s, {{DNNL_ARG_ATTR_SCALES | DNNL_ARG_DST, dst_scales}, ...});
+```
+
+### Option b
+
+Keep data types implicit during scales and zero points setup, but detect the data type from the memory objects passed at execution.
```

Comment:
This option means that we can't pre-generate the simple code for such scales. It will require extending `kernel_` arguments with scales data types and dispatch through GPRs into a proper branch of handling scale code. I don't think it's worth the benefit of saving a single API argument.